### ISSUE TITLE
Add separate scopes for function bodies

### ISSUE THREAD

**Issue by overlookmotel** [oxc-project/backlog#176]
### How it is now

Currently functions only have a single scope which contains bindings for:

* Function name (for function expressions)
* Parameters
* Declarations in body of function

In `SemanticBuilder` we manage to correctly resolve `IdentifierReference`s within function params with an extra call to `resolve_references_for_current_scope` after walking the function params:

https://github.com/oxc-project/oxc/blob/b4407c4e9ae4fd35637c74452eb83997f9494875/crates/oxc_semantic/src/builder.rs#L1955-L1963

i.e. We resolve references in params before any bindings from body of function are added to the scope.

### The problem

In the transformer, we also need to resolve references, but we don't have the advantage that we do in `SemanticBuilder` of constructing `ScopeTree` in order. It already exists, and we have to make sense of it.

Here's an example which doesn't work at present:

```jsx
import React from 'react';
function foo(elem = <Component />) {
    let React;
}
```

In classic mode, JSX transform converts this to:

```js
import React from 'react';
function foo(elem = React.createElement(Component, null)) {
    let React;
}
```

We need to resolve the binding for the inserted reference `React`. Currently, it resolves incorrectly to `let React` inside the function body, as resolution goes up the scope tree and finds a binding for `React` in the function's scope.

### Additional problem

In sloppy mode code, there's also an edge case where you can have 2 bindings for `arguments` in a function - oxc-project/oxc#4232.

### Obvious solution

The obvious solution is to add an additional scope to all function bodies.

Advantage: This is a simple solution which has no "surprises".

Disadvantages:

* More scopes = likely perf hit when resolving references for which the binding is outside current function.
* Makes it more complicated to detect illegal redeclarations (`function foo(x) { let x; }`).

### Alternative solution

* Add `SymbolFlags::Parameter` (or maybe existing `ScopeFlags::FunctionScopedVariable` will do).
* Track in `Traverse` if you're inside a function's parameters or not.
* When resolving a reference, if inside a function's parameters, for the first scope encountered with `ScopeFlags::Function` flag, do not match bindings unless they have a `SymbolFlags::Parameter` flag.

Actually, it's a bit more complex than that, in order to handle cases like this:

```js
const x = 1;
function foo(
    y = function bar(z = x) {
        const x = 3;
    }
) {
  const x = 2;
}
```

The `x` in `z = x` needs to ignore both the binding in `bar` *and* the binding in `foo` and resolve to top level `const x = 1`. So rather than an "in function params" yes/no flag, would need a "function parameter depth" counter (in this case depth is 2).

Advantage: Maintain lower number of scopes = probably better perf.

Disadvantage: The complexity of reference resolution could make it error-prone.

### Which?

The more complicated solution should work (though it wouldn't solve the 2 x `arguments` bindings edge case). But it's somewhat complex.

Personally I would side with trying the simple solution first and measure how much it impacts performance. Perhaps it's negligible. If so, great, but if not then we can look at the more complicated solution.

Any thoughts?

**Comment by overlookmotel** [comment]
@Boshen @Dunqing would appreciate your input on this when you have time.

**Comment by overlookmotel** [comment]
Actually the complex version would need to be more complicated than I thought, to deal with:

```js
const x = 11, y = 12;
function foo(
    y = function bar() {
        const x = 31;
        const xx = x, yy = y; // 31, 12
    }
) {
  const x = 21, y = 22;
}
```

When resolving `x` and `y` in `xx = x, yy = y`, you have to:

* Include bindings in function body of `bar` (to correctly resolve `x`).
* Exclude bindings in function body of `foo` (to correctly resolve `y`).

So it'd need something more complicated than a "function parameter depth" counter. It'd need a stack of some kind to record an "am I in params?" flag for each function you're currently in. Still doable, but a bit complicated.

**Comment by Boshen** [comment]
You are going to hate me for showing you the following foreign language:

In https://tc39.es/ecma262/#sec-functiondeclarationinstantiation:

4. Let formals be func.[[FormalParameters]].

8. Let hasParameterExpressions be [ContainsExpression](https://tc39.es/ecma262/#sec-static-semantics-containsexpression) of formals.

19. If strict is true or hasParameterExpressions is false, then
20. Else,
a. NOTE: A separate [Environment Record](https://tc39.es/ecma262/#sec-environment-records) is needed to ensure that bindings created by [direct eval](https://tc39.es/ecma262/#sec-function-calls-runtime-semantics-evaluation) calls in the formal parameter list are outside the environment where parameters are declared.
b. Let calleeEnv be the LexicalEnvironment of calleeContext.
c. Let env be [NewDeclarativeEnvironment](https://tc39.es/ecma262/#sec-newdeclarativeenvironment)(calleeEnv).
d. [Assert](https://tc39.es/ecma262/#assert): The VariableEnvironment of calleeContext and calleeEnv are the same [Environment Record](https://tc39.es/ecma262/#sec-environment-records).
e. Set the LexicalEnvironment of calleeContext to env.


tldr we need a new scope for binding initializers, this scope points to the parent scope.

**Comment by overlookmotel** [comment]
Original issue https://github.com/oxc-project/oxc/issues/5017 contained a few more comments which GitHub has failed to transfer over to this backlog issue. See original issue for the full conversation.

**Comment by Dunqing** [comment]
A few open issues appear to be related to this proposal:

- https://github.com/oxc-project/oxc/issues/20427
- https://github.com/oxc-project/oxc/issues/22783

These two describe essentially the same problem: references in parameter initializers and function signatures can incorrectly resolve to declarations inside the function body.

There are also downstream examples in Rolldown:

- https://github.com/rolldown/rolldown/issues/10722
- https://github.com/rolldown/rolldown/issues/10779

In #10722, a nested function’s default parameter must not see declarations from its own body, but it must still be able to see a declaration in an enclosing function body—even when that declaration appears later in the source. Resolving it incorrectly causes Rolldown’s dead-code elimination to remove a variable that is still needed.

In #10779, a function-body var incorrectly shadows the same name in a default parameter. Rolldown’s transform.define therefore treats the reference as locally bound and skips the configured replacement.

The common issue is that parameter and body visibility cannot be represented accurately when they share one scope. Giving the function body its own scope would make this boundary explicit and should avoid the increasingly complicated special cases in reference resolution.
